File: php/phpfind/src/phpfind/FindOptions.php
<?php

declare(strict_types=1);

namespace phpfind;

require_once __DIR__ . '/../autoload.php';
//include_once __DIR__ . '/common.php';

/**
 * Class FindOptions
 */
class FindOptions
{
    /**
     * @var FindOption[] $options
     */
    private array $options;
    /**
     * @var array<string, \Closure> $bool_action_map
     */
    private readonly array $bool_action_map;
    /**
     * @var array<string, \Closure> $str_action_map
     */
    private readonly array $str_action_map;
    /**
     * @var array<string, \Closure> $int_action_map
     */
    private readonly array $int_action_map;
    /**
     * @var array<string, string> $bool_map
     */
    private array $bool_map;
    /**
     * @var array<string, string> $str_map
     */
    private array $str_map;
    /**
     * @var array<string, string> $int_map
     */
    private array $int_map;
    /**
     * @var ArgTokenizer $arg_tokenizer
     */
    private ArgTokenizer $arg_tokenizer;

    /**
     * @throws FindException
     */
    public function __construct()
    {
        $this->options = [];

        $this->bool_action_map = [
            'archivesonly' => fn(bool $b, FindSettings $fs) => $fs->set_archives_only($b),
            'colorize' => fn(bool $b, FindSettings $fs) => $fs->colorize = $b,
            'debug' => fn(bool $b, FindSettings $fs) => $fs->set_debug($b),
            'followsymlinks' => fn(bool $b, FindSettings $fs) => $fs->follow_symlinks = $b,
            'excludearchives' => fn(bool $b, FindSettings $fs) => $fs->include_archives = !$b,
            'excludehidden' => fn(bool $b, FindSettings $fs) => $fs->include_hidden = !$b,
            'help' => fn(bool $b, FindSettings $fs) => $fs->print_usage = $b,
            'includearchives' => fn(bool $b, FindSettings $fs) => $fs->include_archives = $b,
            'includehidden' => fn(bool $b, FindSettings $fs) => $fs->include_hidden = $b,
            'nocolorize' => fn(bool $b, FindSettings $fs) => $fs->colorize = !$b,
            'nofollowsymlinks' => fn(bool $b, FindSettings $fs) => $fs->follow_symlinks = !$b,
            'noprintdirs' => fn(bool $b, FindSettings $fs) => $fs->print_dirs = !$b,
            'noprintfiles' => fn(bool $b, FindSettings $fs) => $fs->print_files = !$b,
            'norecursive' => fn(bool $b, FindSettings $fs) => $fs->recursive = !$b,
            'printdirs' => fn(bool $b, FindSettings $fs) => $fs->print_dirs = $b,
            'printfiles' => fn(bool $b, FindSettings $fs) => $fs->print_files = $b,
            'recursive' => fn(bool $b, FindSettings $fs) => $fs->recursive = $b,
            'sort-ascending' => fn(bool $b, FindSettings $fs) => $fs->sort_descending = !$b,
            'sort-caseinsensitive' => fn(bool $b, FindSettings $fs) => $fs->sort_case_insensitive = $b,
            'sort-casesensitive' => fn(bool $b, FindSettings $fs) => $fs->sort_case_insensitive = !$b,
            'sort-descending' => fn(bool $b, FindSettings $fs) => $fs->sort_descending = $b,
            'verbose' => fn(bool $b, FindSettings $fs) => $fs->verbose = $b,
            'version' => fn(bool $b, FindSettings $fs) => $fs->print_version = $b,
        ];

        $this->str_action_map = [
            'in-archiveext' => fn(string $s, FindSettings $fs) => $fs->add_exts($s, $fs->in_archive_extensions),
            'in-archivefilepattern' =>
                fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->in_archive_file_patterns),
            'in-dirpattern' => fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->in_dir_patterns),
            'in-ext' => fn(string $s, FindSettings $fs) => $fs->add_exts($s, $fs->in_extensions),
            'in-filepattern' =>
                fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->in_file_patterns),
            'in-filetype' => fn(string $s, FindSettings $fs) => $fs->add_file_types($s, $fs->in_file_types),
            'maxlastmod' => fn(string $s, FindSettings $fs) => $fs->max_last_mod = new \DateTime($s),
            'minlastmod' => fn(string $s, FindSettings $fs) => $fs->min_last_mod = new \DateTime($s),
            'out-archiveext' => fn(string $s, FindSettings $fs) => $fs->add_exts($s, $fs->out_archive_extensions),
            'out-archivefilepattern' =>
                fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->out_archive_file_patterns),
            'out-dirpattern' =>
                fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->out_dir_patterns),
            'out-ext' => fn(string $s, FindSettings $fs) => $fs->add_exts($s, $fs->out_extensions),
            'out-filepattern' =>
                fn(string $s, FindSettings $fs) => $fs->add_patterns($s, $fs->out_file_patterns),
            'out-filetype' => fn(string $s, FindSettings $fs) => $fs->add_file_types($s, $fs->out_file_types),
            'path' => fn(string $s, FindSettings $fs) => $fs->paths[] = $s,
            'settings-file' => fn(string $s, FindSettings $fs) => $this->update_settings_from_file($fs, $s),
            'sort-by' => fn(string $s, FindSettings $fs) => $fs->set_sort_by($s),
        ];

        $this->int_action_map = [
            'maxdepth' => fn(int $i, FindSettings $fs) => $fs->max_depth = $i,
            'maxsize' => fn(int $i, FindSettings $fs) => $fs->max_size = $i,
            'mindepth' => fn(int $i, FindSettings $fs) => $fs->min_depth = $i,
            'minsize' => fn(int $i, FindSettings $fs) => $fs->min_size = $i,
        ];
        $this->str_map = ['path' => 'path'];
        $this->set_options_from_json();
        $this->arg_tokenizer = new ArgTokenizer($this->bool_map, $this->str_map, $this->int_map);
    }

    private function add_option(FindOption $opt): void
    {
        $this->options[] = $opt;
        if (array_key_exists($opt->long_arg, $this->bool_action_map)) {
            $this->bool_map[$opt->long_arg] = $opt->long_arg;
            if ($opt->short_arg != '') {
                $this->bool_map[$opt->short_arg] = $opt->long_arg;
            }
        } elseif (array_key_exists($opt->long_arg, $this->str_action_map)) {
            $this->str_map[$opt->long_arg] = $opt->long_arg;
            if ($opt->short_arg != '') {
                $this->str_map[$opt->short_arg] = $opt->long_arg;
            }
        } elseif (array_key_exists($opt->long_arg, $this->int_action_map)) {
            $this->int_map[$opt->long_arg] = $opt->long_arg;
            if ($opt->short_arg != '') {
                $this->int_map[$opt->short_arg] = $opt->long_arg;
            }
        }
    }

    /**
     * @throws FindException
     */
    private function set_options_from_json(): void
    {
        $find_options_path = FileUtil::expand_path(Config::FIND_OPTIONS_PATH);
        if (file_exists($find_options_path)) {
            $contents = file_get_contents($find_options_path);
            if ($contents === false || trim($contents) === '') {
                return;
            }
            try {
                $json_obj = (array)json_decode(trim($contents), true, 512, JSON_THROW_ON_ERROR);
                $find_options = $json_obj['findoptions'];
                if ($find_options) {
                    foreach ((array)$find_options as $fo) {
                        $fo = (array)$fo;
                        $short = '';
                        $long = (string)$fo['long'];
                        $desc = (string)$fo['desc'];
                        if (array_key_exists('short', $fo)) {
                            $short = (string)$fo['short'];
                        }
                        $this->add_option(new FindOption($short, $long, $desc));
                    }
                    usort($this->options, array('phpfind\FindOptions', 'cmp_find_options'));
                }
            } catch (\JsonException $e) {
                throw new FindException($e->getMessage());
            }
        } else {
            throw new FindException('File not found: ' . $find_options_path);
        }
    }

    /**
     * @param FindSettings $settings
     * @param array $arg_tokens
     * @return void
     * @throws FindException
     */
    private function update_settings_from_arg_tokens(FindSettings $settings, array $arg_tokens): void
    {
        foreach ($arg_tokens as $arg_token) {
            if ($arg_token->type == ArgTokenType::Bool) {
                if (is_bool($arg_token->value)) {
                    $this->bool_action_map[$arg_token->name]($arg_token->value, $settings);
                    if (in_array($arg_token->name, array("help", "version"))) {
                        return;
                    }
                } else {
                    throw new FindException("Invalid value for option: $arg_token->name");
                }
            } elseif ($arg_token->type == ArgTokenType::Str) {
                if (is_array($arg_token->value)) {
                    foreach ($arg_token->value as $val) {
                        if (is_string($val)) {
                            $this->str_action_map[$arg_token->name]($val, $settings);
                        } else {
                            throw new FindException("Invalid value for option: $arg_token->name");
                        }
                    }
                } elseif (is_string($arg_token->value)) {
                    $this->str_action_map[$arg_token->name]($arg_token->value, $settings);
                } else {
                    throw new FindException("Invalid value for option: $arg_token->name");
                }
            } elseif ($arg_token->type == ArgTokenType::Int) {
                if (is_int($arg_token->value)) {
                    $this->int_action_map[$arg_token->name]($arg_token->value, $settings);
                } else {
                    throw new FindException("Invalid value for option: $arg_token->name");
                }
            } else {
                throw new FindException("Invalid option: $arg_token->name");
            }
        }
    }

    /**
     * @param string $json
     * @param FindSettings $settings
     * @return void
     * @throws FindException
     */
    public function update_settings_from_json(FindSettings $settings, string $json): void
    {
        $arg_tokens = $this->arg_tokenizer->tokenize_json($json);
        $this->update_settings_from_arg_tokens($settings, $arg_tokens);
    }

    /**
     * @param FindSettings $settings
     * @param string $file_path
     * @return void
     * @throws FindException
     */
    private function update_settings_from_file(FindSettings $settings, string $file_path): void
    {
        $arg_tokens = $this->arg_tokenizer->tokenize_file($file_path);
        $this->update_settings_from_arg_tokens($settings, $arg_tokens);
    }

    /**
     * @param FindSettings $settings
     * @param string[] $args
     * @return void
     * @throws FindException
     */
    public function update_settings_from_args(FindSettings $settings, array $args): void
    {
        $arg_tokens = $this->arg_tokenizer->tokenize_args($args);
        $this->update_settings_from_arg_tokens($settings, $arg_tokens);
    }

    /**
     * @param string[] $args
     * @return FindSettings
     * @throws FindException
     */
    public function settings_from_args(array $args): FindSettings
    {
        $settings = new FindSettings();
        // default print_files to true since running as cli
        $settings->print_files = true;
        $this->update_settings_from_args($settings, $args);
        return $settings;
    }

    /**
     * @param FindOption $o1
     * @param FindOption $o2
     * @return int
     */
    private static function cmp_find_options(FindOption $o1, FindOption $o2): int
    {
        return strcmp($o1->sort_arg, $o2->sort_arg);
    }

    /**
     * @return string
     */
    private function get_usage_string(): string
    {
        $usage = "Usage:\n phpfind [options] <path> [<path> ...]";
        $usage .= "\n\nOptions:\n";
        $opt_map = [];
        $longest = 0;
        foreach ($this->options as $option) {
            $opt_str = '';
            if ($option->short_arg) {
                $opt_str = '-' . $option->short_arg . ',';
            }
            $opt_str .= '--' . $option->long_arg;
            if (strlen($opt_str) > $longest) {
                $longest = strlen($opt_str);
            }
            $opt_map[$opt_str] = $option->desc;
        }
        $format_str = " %-" . $longest . "s  %s\n";
        foreach ($opt_map as $opt => $desc) {
            $usage .= sprintf($format_str, $opt, $desc);
        }
        return $usage;
    }

    /**
     * @return void
     */
    public function usage(): void
    {
        echo $this->get_usage_string() . "\n";
    }

    /**
     * @param int $exit_code
     * @return never
     */
    public function usage_and_exit(int $exit_code = 0): never
    {
        $this->usage();
        exit($exit_code);
    }
}

File: haskell/hsfind/src/HsFind/FindOptions.hs
{-# LANGUAGE DeriveGeneric, NoMonomorphismRestriction #-}
module HsFind.FindOptions (
    FindOption(..)
  , FindOptions(..)
  , getFindOptions
  , getUsage
  , parseDateToUtc
  , settingsFromArgs
  , settingsFromFile
  , settingsFromJson) where

import Data.Aeson
import qualified Data.ByteString.Lazy.Char8 as BC
import Data.Char (toLower)
import Data.Either (isLeft, lefts, rights, either)
import Data.List (isPrefixOf, isSuffixOf, sortBy)
import Data.Maybe (fromJust, isJust, listToMaybe)
import Data.Time (UTCTime(..), parseTimeM, defaultTimeLocale)
import GHC.Generics

import HsFind.Paths_hsfind (getDataFileName)
import HsFind.ArgTokenizer
import HsFind.FileTypes (getFileTypeForName)
import HsFind.FileUtil (expandPath, getFileString, isFile)
import HsFind.FindSettings


data FindOption = FindOption
  { long :: String
  , short :: Maybe String
  , desc :: String
  } deriving (Show, Eq, Generic)

instance FromJSON FindOption

newtype JsonFindOptions
  = JsonFindOptions {findoptions :: [FindOption]}
  deriving (Show, Eq, Generic)

instance FromJSON JsonFindOptions

data FindOptions = FindOptions
  { options :: [FindOption]
  , argTokenizer :: ArgTokenizer
  } deriving (Show, Eq, Generic)

findOptionsFile :: FilePath
findOptionsFile = "findoptions.json"

getArgTokenizer :: [FindOption] -> ArgTokenizer
getArgTokenizer jsonOpts =
  ArgTokenizer
    { boolMap = boolMap
    , stringMap = stringMap
    , intMap = intMap
    }
  where
    boolMap = boolShortMap ++ boolLongMap
    stringMap = stringShortMap ++ stringLongMap ++ [("path", "path"), ("settings-file", "settings-file")]
    intMap = intShortMap ++ intLongMap
    boolLongMap :: [(String, String)]
    boolLongMap = [(long o, long o) | o <- jsonOpts, isBoolOption o]
    boolShortMap :: [(String, String)]
    boolShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isBoolOption o, isJust (short o)]
    stringLongMap :: [(String, String)]
    stringLongMap = [(long o, long o) | o <- jsonOpts, isStringOption o]
    stringShortMap :: [(String, String)]
    stringShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isStringOption o, isJust (short o)]
    intLongMap :: [(String, String)]
    intLongMap = [(long o, long o) | o <- jsonOpts, isIntOption o]
    intShortMap :: [(String, String)]
    intShortMap = [(fromJust (short o), long o) | o <- jsonOpts, isIntOption o, isJust (short o)]
    isBoolOption :: FindOption -> Bool
    isBoolOption o = long o `elem` map fst boolActions
    isStringOption :: FindOption -> Bool
    isStringOption o = long o `elem` map fst stringActions
    isIntOption :: FindOption -> Bool
    isIntOption o = long o `elem` map fst integerActions

getFindOptions :: IO (Either String FindOptions)
getFindOptions = do
  findOptionsPath <- getDataFileName findOptionsFile
  findOptionsJsonString <- getFileString findOptionsPath
  case findOptionsJsonString of
    Left e -> return $ Left e
    Right jsonString ->
      case (eitherDecode (BC.pack jsonString) :: Either String JsonFindOptions) of
        Left e -> return $ Left e
        Right jsonFindOptions -> return $ Right FindOptions { options = findoptions jsonFindOptions,
                                                              argTokenizer = getArgTokenizer (findoptions jsonFindOptions) }

getUsage :: FindOptions -> String
getUsage findOptions =
  "Usage:\n hsfind [options] <path> [<path> ...]\n\nOptions:\n" ++ findOptionsToString (options findOptions)

getOptStrings :: [FindOption] -> [String]
getOptStrings = map formatOpts
  where formatOpts FindOption {long=l, short=Nothing} = getLong l
        formatOpts FindOption {long=l, short=Just s}  = shortAndLong s l
        getLong l = "--" ++ l
        shortAndLong s l = "-" ++ s ++ "," ++ getLong l

getOptDesc :: FindOption -> String
getOptDesc FindOption {desc=""} = error "No description for FindOption"
getOptDesc FindOption {desc=d} = d

sortFindOption :: FindOption -> FindOption -> Ordering
sortFindOption FindOption {long=l1, short=s1} FindOption {long=l2, short=s2} =
  compare (shortOrLong s1 l1) (shortOrLong s2 l2)
  where
    shortOrLong Nothing l = l
    shortOrLong (Just s) l = map toLower s ++ "@" ++ l

sortFindOptions :: [FindOption] -> [FindOption]
sortFindOptions = sortBy sortFindOption

padString :: String -> Int -> String
padString s len | length s < len = s ++ replicate (len - length s) ' '
                | otherwise      = s

findOptionsToString :: [FindOption] -> String
findOptionsToString findOptions = 
  unlines $ zipWith formatOptLine optStrings optDescs
  where
    sorted = sortFindOptions findOptions
    optStrings = getOptStrings sorted
    optDescs = map getOptDesc sorted
    longest = maximum $ map length optStrings
    formatOptLine o d = " " ++ padString o longest ++ "  " ++ d

parseDateToUtc :: String -> Maybe UTCTime
parseDateToUtc dateString = 
  parseTimeM True defaultTimeLocale "%Y-%-m-%-d" dateString :: Maybe UTCTime

data ActionType = BoolActionType
                | StringActionType
                | IntegerActionType
                | UnknownActionType
  deriving (Show, Eq)

type BoolAction = FindSettings -> Bool -> FindSettings
type StringAction = FindSettings -> String -> FindSettings
type IntegerAction = FindSettings -> Integer -> FindSettings

boolActions :: [(String, BoolAction)]
boolActions = [ ("archivesonly", \ss b -> ss {archivesOnly=b, includeArchives=b})
              , ("colorize", \ss b -> ss {colorize=b})
              , ("debug", \ss b -> ss {debug=b, verbose=b})
              , ("excludearchives", \ss b -> ss {includeArchives=not b})
              , ("excludehidden", \ss b -> ss {includeHidden=not b})
              , ("followsymlinks", \ss b -> ss {followSymlinks=b})
              , ("help", \ss b -> ss {printUsage=b})
              , ("includearchives", \ss b -> ss {includeArchives=b})
              , ("includehidden", \ss b -> ss {includeHidden=b})
              , ("nocolorize", \ss b -> ss {colorize=not b})
              , ("nofollowsymlinks", \ss b -> ss {followSymlinks=not b})
              , ("noprintdirs", \ss b -> ss {printDirs=b})
              , ("noprintfiles", \ss b -> ss {printFiles=not b})
              , ("norecursive", \ss b -> ss {recursive=not b})
              , ("printdirs", \ss b -> ss {printDirs=b})
              , ("printfiles", \ss b -> ss {printFiles=b})
              , ("recursive", \ss b -> ss {recursive=b})
              , ("sort-ascending", \ss b -> ss {sortDescending=not b})
              , ("sort-caseinsensitive", \ss b -> ss {sortCaseInsensitive=b})
              , ("sort-casesensitive", \ss b -> ss {sortCaseInsensitive=not b})
              , ("sort-descending", \ss b -> ss {sortDescending=b})
              , ("verbose", \ss b -> ss {verbose=b})
              , ("version", \ss b -> ss {printVersion=b})
              ]

stringActions :: [(String, StringAction)]
stringActions = [ ("in-archiveext", \ss s -> ss {inArchiveExtensions = inArchiveExtensions ss ++ newExtensions s})
                , ("in-archivefilepattern", \ss s -> ss {inArchiveFilePatterns = inArchiveFilePatterns ss ++ [s]})
                , ("in-dirpattern", \ss s -> ss {inDirPatterns = inDirPatterns ss ++ [s]})
                , ("in-ext", \ss s -> ss {inExtensions = inExtensions ss ++ newExtensions s})
                , ("in-filepattern", \ss s -> ss {inFilePatterns = inFilePatterns ss ++ [s]})
                , ("in-filetype", \ss s -> ss {inFileTypes = inFileTypes ss ++ [getFileTypeForName s]})
                , ("maxlastmod", \ss s -> ss {maxLastMod = parseDateToUtc s})
                , ("minlastmod", \ss s -> ss {minLastMod = parseDateToUtc s})
                , ("out-archiveext", \ss s -> ss {outArchiveExtensions = outArchiveExtensions ss ++ newExtensions s})
                , ("out-archivefilepattern", \ss s -> ss {outArchiveFilePatterns = outArchiveFilePatterns ss ++ [s]})
                , ("out-dirpattern", \ss s -> ss {outDirPatterns = outDirPatterns ss ++ [s]})
                , ("out-ext", \ss s -> ss {outExtensions = outExtensions ss ++ newExtensions s})
                , ("out-filepattern", \ss s -> ss {outFilePatterns = outFilePatterns ss ++ [s]})
                , ("out-filetype", \ss s -> ss {outFileTypes = outFileTypes ss ++ [getFileTypeForName s]})
                , ("path", \ss s -> ss {paths = paths ss ++ [s]})
                , ("sort-by", \ss s -> ss {sortResultsBy = getSortByForName s})
                ]

integerActions :: [(String, IntegerAction)]
integerActions = [ ("maxdepth", \ss i -> ss {maxDepth = i})
                 , ("maxsize", \ss i -> ss {maxSize = i})
                 , ("mindepth", \ss i -> ss {minDepth = i})
                 , ("minsize", \ss i -> ss {minSize = i})
                 ]

updateSettingsFromTokens :: FindSettings -> FindOptions -> [ArgToken] -> IO (Either String FindSettings)
updateSettingsFromTokens settings findOptions tokens = do
  case tokens of
    [] -> return $ Right settings
    t:ts ->
      case getActionType (name t) of
        BoolActionType ->
          case value t of
            TypeA b -> updateSettingsFromTokens (getBoolAction (name t) settings b) findOptions ts
            _ -> return $ Left $ "Invalid boolean value for option: " ++ name t
        StringActionType ->
          case value t of
            TypeB s ->
              if name t == "settings-file"
              then do
                settingsEither <- updateSettingsFromFile settings findOptions s
                case settingsEither of
                  Left e -> return $ Left e
                  Right settings' -> updateSettingsFromTokens settings' findOptions ts
              else updateSettingsFromTokens (getStringAction (name t) settings s) findOptions ts
            _ -> return $ Left $ "Invalid string value for option: " ++ name t
        IntegerActionType ->
          case value t of
            TypeC i -> updateSettingsFromTokens (getIntegerAction (name t) settings (toInteger i)) findOptions ts
            _ -> return $ Left $ "Invalid integer value for option: " ++ name t
        UnknownActionType -> return $ Left $ "Invalid option from updateSettingsFromTokens: " ++ name t
  where
    getActionType :: String -> ActionType
    getActionType a
      | isBoolAction a = BoolActionType
      | isStringAction a = StringActionType
      | a == "settings-file" = StringActionType
      | isIntegerAction a = IntegerActionType
      | otherwise = UnknownActionType
    getBoolAction :: String -> BoolAction
    getBoolAction a = snd $ head $ filter (\(x,_) -> a==x) boolActions
    getStringAction :: String -> StringAction
    getStringAction a = snd $ head $ filter (\(x,_) -> a==x) stringActions
    getIntegerAction :: String -> IntegerAction
    getIntegerAction a = snd $ head $ filter (\(x,_) -> a==x) integerActions
    isBoolAction :: String -> Bool
    isBoolAction a = isJust $ lookup a boolActions
    isStringAction :: String -> Bool
    isStringAction a = isJust $ lookup a stringActions
    isIntegerAction :: String -> Bool
    isIntegerAction a = isJust $ lookup a integerActions

updateSettingsFromJson :: FindSettings -> FindOptions -> String -> IO (Either String FindSettings)
updateSettingsFromJson settings findOptions jsonStr = do
  let eitherTokens = tokenizeJson (argTokenizer findOptions) jsonStr
  case eitherTokens of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings findOptions tokens

settingsFromJson :: FindOptions -> String -> IO (Either String FindSettings)
settingsFromJson = updateSettingsFromJson defaultFindSettings

updateSettingsFromFile :: FindSettings -> FindOptions -> FilePath -> IO (Either String FindSettings)
updateSettingsFromFile settings findOptions filePath = do
  eitherTokens <- tokenizeFile (argTokenizer findOptions) filePath
  case eitherTokens of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings findOptions tokens

settingsFromFile :: FindOptions -> FilePath -> IO (Either String FindSettings)
settingsFromFile = updateSettingsFromFile defaultFindSettings

updateSettingsFromArgs :: FindSettings -> FindOptions -> [String] -> IO (Either String FindSettings)
updateSettingsFromArgs settings findOptions arguments = do
  case tokenizeArgs (argTokenizer findOptions) arguments of
    Left e -> return $ Left e
    Right tokens -> updateSettingsFromTokens settings findOptions tokens

settingsFromArgs :: FindOptions -> [String] -> IO (Either String FindSettings)
settingsFromArgs = updateSettingsFromArgs defaultFindSettings{printFiles=True}
